bittensor/_receptor/receptor_pool_impl.py

```diff
@@ -133,28 +191,109 @@ def forward (
+
+            future_map = {}
+            for idx, call_arg in enumerate(call_args):
+                future = executor.submit( call_forward, call_arg)
+                future_map[future] = call_arg
+
+            success_response_cnt = 0 # success count 
+            for i,future in enumerate(concurrent.futures.as_completed(future_map)):
+                
+                # get future_call_argds
+                future_call_args = future_map.pop(future)
+
+                # get uid (for metagraph retreival) make 1 uid per synapse (same uid)
+                endpoint_uid = [future_call_args['receptor'].endpoint.uid ]*len(synapses)
+                
+                # get respone
+                response = future.result()
+
+                if response[1][0] == 1:
+                    # this indicates a successful response 
+                    success_response_cnt += 1
+                    results_dict['outputs'].append( response[0] )
+                    results_dict['codes'].append( response[1] )
+                    results_dict['times'].append( response[2] )
+                    results_dict['uids'].append(endpoint_uid)
+                else:
+                    # when return_success_only, ignore the responses that arent successful        
+                    if not return_success_only:
+                        results_dict['outputs'].append( response[0] )
+                        results_dict['codes'].append( response[1] )
+                        results_dict['times'].append( response[2] )
+                        results_dict['uids'].append(endpoint_uid)
+
+                future_call_args['receptor'].semaphore.release()  
+
+                if graph != None:
+                    graph_state_dict = graph.state_dict()
+                    for k in graph_features:
+                        results_dict[k] = [[v]*len(synapses) for v in graph_state_dict[k][[i[0] for i in results_dict['uids']]].tolist()]
+        
+                # when the success_response_cnt > min_success
+                if success_response_cnt >= min_success:
+                    break
+
+        # cancel all the running futures and delete the future map
+        for future, future_call_args in future_map.items():
+            future.cancel()
+            future_call_args['receptor'].semaphore.release()
+        del future_map
+ 
         self._destroy_receptors_over_max_allowed()
-        # ---- Return ----
-        return forward_outputs, forward_codes, forward_times
+
+
+
+        # resolve output type as either a tuple or a dictionary
+
+        if return_type in ['tuple', tuple]:
+            return_result =  [results_dict['outputs'], 
+                    results_dict['codes'], 
+                    results_dict['times'],
+                    results_dict['uids']]
+
+            # add graph features if they exists
+            for k in graph_features:
+                if k in results_dict:
+                    return_result.append(results_dict[k])
```

Comment:
I don't think we should return graph features if `return_type == tuple`, and I think we should leave it as the default (legacy) setting and try to push the `dict` return type whenever we can (refactoring the code to do so now might be beyond the scope of this PR).

My reasoning is that.. well firstly, I don't like getting 3 different tensors back and having to figure out what they are. String labels are way better. 

Secondly and more importantly, what if we change the default value for `graph_features`? If the length changes, then we'll (at best) break any call to this function that expects an exact # of values returned even if they don't explicitly request it. We're actually seeing this in the unit tests because the `uids` are returned in the tuple now and the tests are expecting 3 values when you're handing them 4. At worst we'd raise the more pernicious

More perniciously, we could change the order of items in the default arg, get them in the wrong order, and truly ruin someones day as they try to debug why their previously-working code is throwing tantrums for no real reason.

the dict approach is just all-around better for legibility and safety.

Reply:
Also, consider the return type hinting for this function. Currently it's `Tuple[List[torch.Tensor], List[int], List[float]]`. With the graph features it'd be something like `Tuple[List[torch.Tensor], List[int], List[float], List[int], , List[float], List[float], List[float], List[float], List[float], List[float], List[float]]`. Unless you change the number of metagraph features to return.

Compared to the `Dict[str, Union[torch.Tensor, int, float]]` which is just 🔥🔥🔥

Reply:
so we can fix the metagraph features, but yah, i think we should include it in the new features haha.